### PR TITLE
[issue-46] Merge opossum datastructures

### LINKED ISSUE #46: Add functionality to merge two opossum datastructures
The goal of this ticket is to create a function that receives two opossum data structures and returns a merged version of these. The data structure is defined in OpossumUI in `src/ElectronBackend/input/OpossumInputFileSchema.json`.

**Relevant fields are:**
*     "resources", "metadata", "externalAttributions", "resourcesToAttributions"

**Features:**
* metadata: the first one should be taken
* resources are merged and deduplicated
** if there is a name clash between file and folder, it should raise an error
* attributions are merged and deduplicated
** handle case of same attribution ids in two different files (e.g. add prefix to ids)
* attributions to resources are merged and deduplicated

### PR DESCRIPTION
**Pull request by meretp**
This PR adds a method to merge multiple opossum datastructures. 

In the description of the ticket we said that we want to raise an error if there is a name clash between file and folder when merging resources. So far the implementation doesn't distinguish between file and folder, every leaf in the resource tree is a file. I opened an issue to implement this which should then also take care of the merge functionality (#60).

I decided to implement merge functionalities for all fields currently present in the `OpossumInformation` class. All based on the same principle: merge and deduplicate.
Additionally, I decided to expand the internal IDs for the attributions with the project-id to make them unique per document. As we previously used identical ids for the project I had to adapt the implementation to use UUIDs for the projectID.
 
fixes #46

### CODE REVIEW

File: tests/test_merge.py [hunk 1]
```diff
@@ -0,0 +1,141 @@
...
+    )
+
+
+def test_merge_resources_to_attributions() -> None:
+    resources_to_attributions = [
+        {"resources/Path": ["identifier-A", "identifier-B"]},
+        {"resources/Path": ["identifier-C"]},
+        {"resources/Path/different": ["identifier-A"]},
+    ]
+    resources_to_attribution = _merge_resources_to_attributions(
```
Comment (lennartclaas) [inline tests/test_merge.py:87]: Naming could be a bit confusing here as `resources_to_attribution` indicates that a resource can only be assigned with a single attribution. Maybe something like `merged_resources_to_attributions` is clearer.

File: tests/test_merge.py [hunk 2]
```diff
@@ -0,0 +1,141 @@
...
+
+    assert merged_resource == Resource(
+        {
+            "A": Resource({"B": Resource({"C": Resource({})}), "D": Resource({})}),
+            "C": Resource({"D": Resource({"E": Resource({})})}),
+        }
+    )
+
+
+def test_merge_resources_to_attributions() -> None:
```
Comment (lennartclaas) [inline tests/test_merge.py:81]: Another edge case could be tested: 

```
resources_to_attributions = [
  {'folder/file': ['uuid_1']},
  {'folder/file': ['uuid_1', 'uuid_2']},
]
```

**Review by lennartclaas — CHANGES_REQUESTED** [review]
Looks good. Only some minor points.

File: tests/test_merge.py [hunk 3]
```diff
@@ -0,0 +1,141 @@
...
+    SourceInfo,
+)
+
+
+@mock.patch("opossum_lib.opossum_file.OpossumPackage", autospec=True)
+def test_merge_opossum_information(opossum_package: OpossumPackage) -> None:
+    opossum_information = OpossumInformation(
+        Metadata("project-id", "30-05-2023", "test data"),
+        Resource({"A": Resource({"B": Resource({})})}),
+        {"SPDXRef-Package": opossum_package},
```
[the remark this answers is not in the document]
Comment (meretp) [reply tests/test_merge.py:30]: Opened a ticket: #62.

File: tests/test_merge.py (hunk 1 above)
Comment (meretp) [reply tests/test_merge.py:87]: Good catch, I changed the name accordingly.

File: tests/test_merge.py [hunk 4]
```diff
@@ -0,0 +1,141 @@
...
+        {"/A/B/": ["SPDXRef-Package"]},
+    )
+
+    opossum_information_2 = OpossumInformation(
+        Metadata("test-data-id", "29-05-2023", "second test data"),
+        Resource({"A": Resource({"D": Resource({"C": Resource({})})})}),
+        {"SPDXRef-File": opossum_package},
+        {"/A/D/C/": ["SPDXRef-File"]},
```
[the remark this answers is not in the document]
Comment (meretp) [reply tests/test_merge.py:38]: I created a ticket to distinguish between file and folder (#60) in general and added a comment to take care of the paths. Apart from that I adapted this testcase to be consistent with the definiton in the input schema.

File: tests/test_merge.py (hunk 2 above)
Comment (meretp) [reply tests/test_merge.py]: I added this test case and parametrized the test accordingly, also I used the naming I had in the other test case to be consistent. Once we adapt the other naming conventions this should also be changed.

**Review by lennartclaas — APPROVED** [review]
LGTM. Only a single optional point.

File: tests/test_merge.py [hunk 5]
```diff
@@ -0,0 +1,158 @@
...
+            },
+        ),
+        (
+            [{"resources/Path": ["uuid_1"]}, {"resources/Path": ["uuid_1", "uuid_2"]}],
+            {
+                "resources/Path": ["uuid_1", "uuid_2"],
+            },
+        ),
+    ],
+)
```
[the remark this answers is not in the document]
Comment (meretp) [reply tests/test_merge.py:104]: I agree that this is hard to read but I think it's better to parametrize a test than to duplicate it (accroding to the DRY principle). As this is also just a test and I'd say that readability is not as important as for the code, I will leave it as it is and merge it.
